feather/server/src/systems/gamemode.rs

```diff
@@ -0,0 +1,226 @@
+use std::collections::HashMap;
+
+use base::Gamemode;
+use common::Game;
+use ecs::{SysResult, SystemExecutor};
+use quill_common::components::{
+    CanBuild, CanCreativeFly, CreativeFlying, CreativeFlyingSpeed, Instabreak, Invulnerable,
+    PreviousGamemode, WalkSpeed,
+};
+use quill_common::events::{
+    BuildingAbilityChangeEvent, CreativeFlyingEvent, FlyingAbilityChangeEvent, GamemodeEvent,
+    InstabreakChangeEvent, InvulnerabilityChangeEvent,
+};
+
+use crate::{ClientId, Server};
+use base::anvil::player::PlayerAbilities;
+
+pub fn register(systems: &mut SystemExecutor<Game>) {
+    systems.group::<Server>().add_system(gamemode_change);
+}
+
+fn gamemode_change(game: &mut Game, server: &mut Server) -> SysResult {
+    let mut may_fly_changes = HashMap::new();
```

Comment:
Can be a vector or we could just send the event on change instead?

Reply:
If I remember correctly, I was having trouble with borrow checker because `game` is already mutably borrowed because of query. It can't be a vector because its value may be true or false or nothing (property not changed), and we have to dispatch an event in the first two cases, but not the third.

Reply:
If you extract ecs into a separate var the you should be able to do it. Also a vector of triples should be faster than a hashmap, if you can't get around the borrow checker.